service/storage_proxy.cc

```diff
@@ -1909,14 +1909,18 @@ future<> storage_proxy::mutate_counters(Range&& mutations, db::consistency_level
         // so if there is a timeout we don't really know which one is to "blame"
         // and what to put in ks and cf fields of write timeout exception.
         // Let's just use the schema of the first mutation in a vector.
-        auto handle_error = [this, sp = this->shared_from_this(), s = endpoint_and_mutations.second[0].s, cl, permit] (std::exception_ptr exp) {
+        auto handle_error = [this, sp = this->shared_from_this(), s = endpoint_and_mutations.second[0].s, endpoint, cl, permit] (std::exception_ptr exp) {
             auto& ks = _db.local().find_keyspace(s->ks_name());
             try {
                 std::rethrow_exception(std::move(exp));
             } catch (rpc::timeout_error&) {
                 return make_exception_future<>(mutation_write_timeout_exception(s->ks_name(), s->cf_name(), cl, 0, db::block_for(ks, cl), db::write_type::COUNTER));
             } catch (timed_out_error&) {
                 return make_exception_future<>(mutation_write_timeout_exception(s->ks_name(), s->cf_name(), cl, 0, db::block_for(ks, cl), db::write_type::COUNTER));
+            } catch (rpc::closed_error&) {
+                const auto msg = fmt::format("Node {} closed connection while sending counter mutations", endpoint);
+                slogger.warn(msg.c_str());
+                return make_exception_future<>(unavailable_exception(msg, db::consistency_level::ONE, 1, 0));
```

Comment:
mutation_write_failure_exception exception is more appropriate here IMO. Also report real CL, it would be surprising to see CL=ONE errors on CL=QUORUM requests.

Reply:
- CL=ONE is an idea from https://github.com/scylladb/scylla/blob/390c8b9b3710b482bfda19b0599fd5da48324f85/service/storage_proxy.cc#L2120
After all, this is counter mutation being sent to leader. There is no concept of CL yet, because we have only one leader. In the case of timeout the CL displayed is the real CL of the query (just as you recommend), but isn't it more misleading?

- I tried `mutation_write_failure_exception` but apparently cassandra-stress handles it differently than `unavailable_exception`. C-S becomes defunct in the former case (stops querying after a RuntimeError). In the latter case C-S prints some IOExceptions, but at least keeps querying, so we'd get something "visibly fixed" with such solution.

What do you think about it? In any case, I agree that this choice of exception type is far from perfect.

Reply:
That was a bad idea there too for other reasons but luckily it almost impossible to hit in practice. Unlike this one you are adding. There is no indication in an error at what stage of write handling it happened, so for a client it is no more clear what is going on. Any error on a coordinator may prevent a query from succeeding, but we do not report them as CL=ONE.

About c-s not handling an error may be you are using old c-s that does not recognize the error. Try using the latest one from Cassandra.

Reply:
@gleb-cloudius OK, I changed the exception type (with `block_for()` appropriate for CL and `failures==1`).
Also, I built C-S from apache trunk. The result is that it **sometimes** still hangs on `mutation_write_failure_exception` (error below).

```
java.io.IOException: Operation x10 on key(s) [34344b4f3633374f3031]: Error executing: (WriteFailureException): Cassandra failure during write query at consistency QUORUM (2 responses were required but only 0 replica responded, 1 failed)

	at org.apache.cassandra.stress.Operation.error(Operation.java:127)
	at org.apache.cassandra.stress.Operation.timeWithRetry(Operation.java:105)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:91)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:99)
	at org.apache.cassandra.stress.operations.predefined.CqlOperation.run(CqlOperation.java:248)
	at org.apache.cassandra.stress.StressAction$Consumer.run(StressAction.java:466)
com.datastax.driver.core.exceptions.WriteFailureException: Cassandra failure during write query at consistency QUORUM (2 responses were required but only 0 replica responded, 1 failed)
com.datastax.driver.core.exceptions.WriteFailureException: Cassandra failure during write query at consistency QUORUM (2 responses were required but only 0 replica responded, 1 failed)
com.datastax.driver.core.exceptions.WriteFailureException: Cassandra failure during write query at consistency QUORUM (2 responses were required but only 0 replica responded, 1 failed)
com.datastax.driver.core.exceptions.WriteFailureException: Cassandra failure during write query at consistency QUORUM (2 responses were required but only 0 replica responded, 1 failed)
total,                                                 69121,    5502,    5502,    5502,     2,9,     2,8,     4,8,     6,5,     8,2,     9,7,   16,6, -0,56261,      1,      0,       0,       0,       0,       0
FAILURE
java.lang.RuntimeException: Failed to execute stress action
	at org.apache.cassandra.stress.StressAction.run(StressAction.java:99)
	at org.apache.cassandra.stress.Stress.run(Stress.java:143)
	at org.apache.cassandra.stress.Stress.main(Stress.java:62)
```

Reply:
Sometimes it fails even with too much timeout errors. There is a flag to ignore errors for that.